homeassistant/components/neato/__init__.py

```diff
@@ -32,136 +31,110 @@
     {
         NEATO_DOMAIN: vol.Schema(
             {
-                vol.Required(CONF_USERNAME): cv.string,
-                vol.Required(CONF_PASSWORD): cv.string,
-                vol.Optional(CONF_VENDOR, default="neato"): vol.In(VALID_VENDORS),
+                vol.Required(CONF_CLIENT_ID): cv.string,
+                vol.Required(CONF_CLIENT_SECRET): cv.string,
             }
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
+PLATFORMS = ["camera", "vacuum", "switch", "sensor"]
 
-async def async_setup(hass, config):
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
     """Set up the Neato component."""
+    hass.data[NEATO_DOMAIN] = {}
 
     if NEATO_DOMAIN not in config:
-        # There is an entry and nothing in configuration.yaml
         return True
 
-    entries = hass.config_entries.async_entries(NEATO_DOMAIN)
     hass.data[NEATO_CONFIG] = config[NEATO_DOMAIN]
+    vendor = Neato()
+    config_flow.OAuth2FlowHandler.async_register_implementation(
+        hass,
+        api.NeatoImplementation(
+            hass,
+            NEATO_DOMAIN,
+            config[NEATO_DOMAIN][CONF_CLIENT_ID],
+            config[NEATO_DOMAIN][CONF_CLIENT_SECRET],
+            vendor.auth_endpoint,
+            vendor.token_endpoint,
+        ),
+    )
 
-    if entries:
-        # There is an entry and something in the configuration.yaml
-        entry = entries[0]
-        conf = config[NEATO_DOMAIN]
-        if (
-            entry.data[CONF_USERNAME] == conf[CONF_USERNAME]
-            and entry.data[CONF_PASSWORD] == conf[CONF_PASSWORD]
-            and entry.data[CONF_VENDOR] == conf[CONF_VENDOR]
-        ):
-            # The entry is not outdated
-            return True
-
-        # The entry is outdated
-        error = await hass.async_add_executor_job(
-            NeatoConfigFlow.try_login,
-            conf[CONF_USERNAME],
-            conf[CONF_PASSWORD],
-            conf[CONF_VENDOR],
-        )
-        if error is not None:
-            _LOGGER.error(error)
-            return False
-
-        # Update the entry
-        hass.config_entries.async_update_entry(entry, data=config[NEATO_DOMAIN])
-    else:
-        # Create the new entry
+    return True
+
+
+async def async_migrate_entry(hass: HomeAssistantType, config_entry: ConfigEntry):
```

Comment:
We can remove this function.

Reply:
This results in `ERROR (MainThread) [homeassistant.config_entries] Migration handler not found for entry <entry_name> for neato`

Reply:
Wondering if we should not change the version at all?

Reply:
I really like the version as an indicator whether the entry is updated or not. Making guesses based on properties doesn't really seem reliable to me.

Reply:
We know that the `"token"` key will be missing in the old config entry data. We could use that fact.

Reply:
Done.